sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,93 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
+    val (preprocess, arrayData, arrayWriter) =
+      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
+
+    ev.copy(code =
+      preprocess +
       ctx.splitExpressions(
         ctx.INPUT_ROW,
-        children.zipWithIndex.map { case (e, i) =>
-          val eval = e.genCode(ctx)
-          eval.code + s"""
-            if (${eval.isNull}) {
-              $values[$i] = null;
+        evals.zipWithIndex.map { case (eval, i) =>
+          eval.code +
+            (if (isPrimitiveArray) {
+              (if (!children(i).nullable) {
+                s"\n$arrayWriter.write($i, ${eval.value});"
+              } else {
+                s"""
+                if (${eval.isNull}) {
+                  $arrayWriter.setNull$primitiveTypeName($i);
+                } else {
+                  $arrayWriter.write($i, ${eval.value});
+                }
+               """
+              })
             } else {
-              $values[$i] = ${eval.value};
-            }
-           """
+              s"""
+              if (${eval.isNull}) {
+                $array[$i] = null;
+              } else {
+                $array[$i] = ${eval.value};
+              }
+             """
+            })
         }) +
-      s"""
-        final ArrayData ${ev.value} = new $arrayClass($values);
-        this.$values = null;
-      """, isNull = "false")
+      s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+      isNull = "false")
   }
 
   override def prettyName: String = "array"
 }
 
+private [sql] object GenArrayData {
+  // This function returns Java code pieces based on DataType and isPrimitive
+  // for allocation of ArrayData class
+  def getCodeArrayData(
+      ctx: CodegenContext,
+      dt: DataType,
+      size: Int,
+      isPrimitive : Boolean,
+      array: String): (String, String, String) = {
+    if (!isPrimitive) {
+      val arrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", array,
+        s"this.$array = new Object[${size}];")
+      ("", s"new $arrayClass($array)", null)
+    } else {
+      val holder = ctx.freshName("holder")
+      val arrayWriter = ctx.freshName("createArrayWriter")
+      val unsafeArrayClass = classOf[UnsafeArrayData].getName
+      val holderClass = classOf[BufferHolder].getName
+      val arrayWriterClass = classOf[UnsafeArrayWriter].getName
+      ctx.addMutableState(unsafeArrayClass, array, "")
+      ctx.addMutableState(holderClass, holder, "")
+      ctx.addMutableState(arrayWriterClass, arrayWriter, "")
+      val baseOffset = Platform.BYTE_ARRAY_OFFSET
+      val unsafeArraySizeInBytes =
+        UnsafeArrayData.calculateHeaderPortionInBytes(size) +
+        ByteArrayMethods.roundNumberOfBytesToNearestWord(dt.defaultSize * size)
+
+      (s"""
+        $array = new $unsafeArrayClass();
+        $holder = new $holderClass($unsafeArraySizeInBytes);
+        $arrayWriter = new $arrayWriterClass();
```

Comment:
When  I did these things (more precisely allocation of `UnsafeArrayData`), it caused test failure.
Thus, I intentionally created `UnsafeArrayData` every time when we evaluate the expression with an input row.